python/dgl/graphbolt/impl/torch_based_feature_store.py

```diff
@@ -12,7 +13,168 @@
+
+    Initialize a disk based feature fetcher by a numpy file.
+
+    Parameters
+    ----------
+    path : string
+        The path to the numpy feature file.
+        Note that the dimension of the numpy should be greater than 1.
+
+    Examples
+    --------
+    >>> import torch
+    >>> from dgl import graphbolt as gb
+
+    >>> torch_feat = torch.arange(10).reshape(2, -1)
+    >>> feature = gb.DiskBasedFeature(torch_feat)
+    >>> feature.read(torch.tensor([0]))
+    tensor([[0, 1, 2, 3, 4]])
+    >>> feature.size()
+    torch.Size([5])
+
+    """
+
+    def __init__(self, path: str, metadata: Dict = None):
+        super().__init__()
+        self._is_inplace_pinned = set()
+        mmap_mode = "r+"
+        self._tensor = torch.from_numpy(
+            np.load(path, mmap_mode=mmap_mode)
+        ).contiguous()
+
+        # Disk feature path.
+        self._path = path
+        self._metadata = metadata
+
+    def __del__(self):
+        # torch.Tensor.pin_memory() is not an inplace operation. To make it
+        # truly in-place, we need to use cudaHostRegister. Then, we need to use
+        # cudaHostUnregister to unpin the tensor in the destructor.
+        # https://github.com/pytorch/pytorch/issues/32167#issuecomment-753551842
+        for tensor in self._is_inplace_pinned:
+            assert self._inplace_unpinner(tensor.data_ptr()) == 0
+
+    def read(self, ids: torch.Tensor = None):
+        """Read the feature by index.
+
+        The returned tensor will be on CPU.
+
+        Parameters
+        ----------
+        ids : torch.Tensor
+            The index of the feature. Only the specified indices of the
+            feature are read.
+
+        Returns
+        -------
+        torch.Tensor
+            The read feature.
+        """
+        if ids is None:
+            if self._tensor.is_pinned():
+                return self._tensor.cuda()
+            return self._tensor
+        elif platform.system() == "Linux":
+            try:
+                return torch.ops.graphbolt.disk_index_select(
+                    self._path, ids.to("cpu"), self._tensor.dtype
+                ).to(ids.device)
+            except RuntimeError:
+                raise IndexError
+        else:
+            return index_select(self._tensor, ids)
+
+    def size(self):
+        """Get the size of the feature.
+        Returns
+        -------
+        torch.Size
+            The size of the feature.
+        """
+        return torch.Size(torch.ops.graphbolt.disk_feature_size(self._path))
+
+    def update(self, value: torch.Tensor, ids: torch.Tensor = None):
+        raise NotImplementedError
+
+    def metadata(self):
+        """Get the metadata of the feature.
+
+        Returns
+        -------
+        Dict
+            The metadata of the feature.
+        """
+        return (
+            self._metadata if self._metadata is not None else super().metadata()
+        )
+
+    def pin_memory_(self):
```

Comment:
We are not going to support pinning DiskBasedFeature so we can remove these sections.

Reply:
I do not add it or retain it as NotImplement at first , but there are many tests and examples needs this interface. I will try to integrate within the torchfeature to avoid such problem.

Reply:
Why not use BasicFeatureStore to store the DiskBasedFeatures? Then, we have two feature stores, one is TorchBasedFeatureStore, one BasicFeatureStore if needed.

@frozenbugs What is the best way to do this?